wasm/validator/src/recursive/isi.rs

```diff
@@ -0,0 +1,150 @@
+//! Module with [`Instruction`] sub-sets and utils.
+
+use super::*;
+
+/// Same as [`Instruction`] but split instructions into two sub-sets:
+/// [`DirectInstruction`] and [`IndirectInstruction`].
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum EitherInstruction {
+    /// [`DirectInstruction`] variant.
+    Direct(DirectInstruction),
+    /// [`IndirectInstruction`] variant.
+    Indirect(IndirectInstruction),
+}
+
+/// Instruction without any sub-instructions like [`If`](IfInstruction).
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum DirectInstruction {
+    /// `Register` variant.
+    Register(RegisterBox),
+    /// `Unregister` variant.
+    Unregister(UnregisterBox),
+    /// `Mint` variant.
+    Mint(MintBox),
+    /// `Burn` variant.
+    Burn(BurnBox),
+    /// `Transfer` variant.
+    Transfer(TransferBox),
+    /// `Fail` variant.
+    Fail(FailBox),
+    /// `SetKeyValue` variant.
+    SetKeyValue(SetKeyValueBox),
+    /// `RemoveKeyValue` variant.
+    RemoveKeyValue(RemoveKeyValueBox),
+    /// `Grant` variant.
+    Grant(GrantBox),
+    /// `Revoke` variant.
+    Revoke(RevokeBox),
+    /// `ExecuteTrigger` variant.
+    ExecuteTrigger(ExecuteTriggerBox),
+    /// `SetParameter` variant.
+    SetParameter(SetParameterBox),
+    /// `NewParameter` variant.
+    NewParameter(NewParameterBox),
+    /// `Upgrade` variant.
+    Upgrade(UpgradeBox),
+}
+
+/// Instruction with sub-instructions.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum IndirectInstruction {
+    If(Box<IfInstruction>),
+    Pair(Box<Pair>),
+    Sequence(SequenceBox),
+}
```

Comment:
should we even have these?

Reply:
Couldn't you disassemble them on the host side and feed them to the validator one by one? they can also be batched
The same goes for validating a `SignedTransaction` which is essentially just a collection of instructions

Reply:
That's how we did it before. And I'm not sure it was the best idea.
User could desire to validate `If`, `Pair` and `Sequence` instructions as well. For example deny all `If`s.

What I wanted to do is to make a single entrypoint to a valdiator -- transaction. And then validator decides how to handle it. However in most cases validator needs to do the same stuff, so I extracted it in `iroha_validator`.


But your approach can fix one problem of my approach. With your way we can keep validators as validators. With my way we should rename them into *verifiers* (that's what @appetrosyan suggested)

Reply:
I appreciate that you tried to encode this distinction in the type system but I don't like the partitioning this creates between iroha `data_model` and wasm `data_model`. I've been trying hard to keep the API the same whether the user is writing an app or a smart contract and I would like the uniform API to remain one of our goals. Taking this into account, I would either move this code to `iroha_data_model` or just remove it and make it a runtime thing as was done previously

Reply:
Let's revisit this later, after we finish other discussions

Reply:
as I've said earlier we want a unified API for `iroha_client` and `iroha_wasm`. This means the same dev experience and the same `data_model`

IMO you should ask yourself: "Is this struct I've introduced **only** used in `iroha_wasm`"? If yes, in the level of preference:
1. Try harder to remove it. API can be redefined or we can move from strong typing of invariants to runtime checking
2. If it is likely this struct will be used in other places, not just in `iroha_wasm`, we can consider leaving it in `data_model` as long as there is tracking issue that defines where we'll make use of that struct other than `wasm`
3. is there a tangible benefit of having this struct that would justify it's existence and split in the API. Does it prevent some horrible accidental bugs for the user that cannot be prevented in any other means? I don't see any benefit of strong typing `TransactionInfo` (it looks to be there only to make your life easier, not user's). However, I'm not sure about `DirectInstruction`?

Reply:
Yes, it's used only in `iroha_wasm` (in `itoha_validator` to be precise), that's why I placed it here before.

1. I don't like the idea of runtime invariants then we can use compile-time invariants. With that user can't make a mistake.
3. `TransactionInfo` prevents user to check every instruction mannually, I think it's good.

I still can't see a big problem here. `iroha_client`, `iroha_wasm` and `iroha_validator`  -- all three serve to a three different purposes. If we can do some pre-processing and make API more obvious then let's do that